Added SQL function CRC32

aestriplex opened the pull request:
# Description

this PR implements the CRC32 function discussed in #29856

# Checklist

- [x] Is the user manual updated?
- [ ] Are the test cases passed and automated?
- [x] Is there no significant decrease in test coverage?

dapan1121 commented on docs/en/14-reference/03-taos-sql/10-function.md:
````diff
@@ -1159,6 +1159,61 @@ taos> select repeat('abc',-1);
                   |
 ```
 
+#### CRC32
````
中文文档没有修改

aestriplex replied:
I just modified the English file because I don't know Chinese. If it's ok for you, I'll translate with an automatic translator.

我只是修改了英文文件，因为我不懂中文。如果你觉得可以，我可以用自动翻译器翻译。

aestriplex replied:
@dapan1121 I added the Chinese version too. I double checked it with two translators and with other Chinese specifications. A native speaker should review it though, because, as I said, I don't speak Chinese

我還加了中文版本。我用兩個自動翻譯器和其他中文規格進行了雙重檢查。但應該由母語人士審核，因為正如我之前所說，我不懂中文。

dapan1121 commented on docs/en/14-reference/03-taos-sql/10-function.md:
````diff
@@ -1159,6 +1159,61 @@ taos> select repeat('abc',-1);
+#### CRC32
+
+```sql
+CRC32(string)
````
输入参数应该是表达式

dapan1121 commented on docs/en/14-reference/03-taos-sql/10-function.md:
```diff
@@ -1159,6 +1159,61 @@ taos> select repeat('abc',-1);
+
+**Return Type**: UINT. 
+
+**Applicable Data Types**: VARCHAR, NCHAR.
```
应该允许所有类型

dapan1121 commented on docs/en/14-reference/03-taos-sql/10-function.md:
```diff
@@ -1159,6 +1159,61 @@ taos> select repeat('abc',-1);
+
+**Usage Instructions**:
+
+- If `string` is NULL, it raises the error `Invalid parameter data type`.
```
输入为NULL，输出也应该为NULL

dapan1121 commented on include/libs/function/functionMgt.h:
```diff
@@ -111,6 +111,7 @@ typedef enum EFunctionType {
   FUNCTION_TYPE_REPLACE,
   FUNCTION_TYPE_REPEAT,
   FUNCTION_TYPE_SUBSTR_IDX,
+  FUNCTION_TYPE_CRC32,
```
它应该是数学函数

aestriplex replied:
Ok, I moved the docs to the math functions section too

好的，我也将文档移到了数学函数部分

dapan1121 commented on source/libs/function/src/builtins.c:
```diff
@@ -5129,6 +5135,27 @@ const SBuiltinFuncDefinition funcMgtBuiltins[] = {
+                   .inputParaInfo[0][0] = {.isLastParam = true,
+                                           .startParam = 1,
+                                           .endParam = 1,
+                                           .validDataType = FUNC_PARAM_SUPPORT_VARCHAR_TYPE,
```
类型不足

dapan1121 commented on source/libs/function/src/builtins.c:
```diff
@@ -5129,6 +5135,27 @@ const SBuiltinFuncDefinition funcMgtBuiltins[] = {
+                                           .validNodeType = FUNC_PARAM_SUPPORT_EXPR_NODE,
+                                           .paramAttribute = FUNC_PARAM_NO_SPECIFIC_ATTRIBUTE,
+                                           .valueRangeFlag = FUNC_PARAM_NO_SPECIFIC_VALUE,},
+                   .outputParaInfo = {.validDataType = FUNC_PARAM_SUPPORT_VARCHAR_TYPE}},
```
输出不是UINT吗

dapan1121 commented on source/libs/scalar/src/sclfunc.c:
```diff
@@ -1307,6 +1307,95 @@ int32_t md5Function(SScalarParam* pInput, int32_t inputNum, SScalarParam* pOutpu
+  
+  uint32_t *out = (uint32_t *) pOutputData->pData;
+  for (int32_t i = 0; i < pInput->numOfRows; ++i) {
+    int32_t colId = (pInput->numOfRows == 1) ? 0 : i;
```
colId是不是永远等于i?

aestriplex commented:
@dapan1121 I'm fixing all the comments you reported to me. Should I put them in a separate commit or would you rather I do a rebase?

我正在修复您报告给我的所有评论。我应该将它们放在单独的提交中，还是您希望我进行重新定基？

aestriplex commented:
@dapan1121 I've added a separate commit for cr fixes

我为 cr 修复添加了一个单独的提交

aestriplex commented:
I fixed my tests in test_functions.py. They now pass, so I updated the checkbox in the PR description

dapan1121 commented on docs/zh/14-reference/03-taos-sql/10-function.md:
````diff
@@ -609,6 +609,75 @@ taos> select truncate(8888.88, -1);
+```sql
+CRC32(expr)
+```
+**功能说明**: 傳回代表循環冗餘檢查 (CRC) 的無符號 32 位元整數。
````
这里使用的是繁体中文

dapan1121 commented on source/libs/scalar/src/sclfunc.c:
```diff
@@ -1307,6 +1307,113 @@ int32_t md5Function(SScalarParam* pInput, int32_t inputNum, SScalarParam* pOutpu
+  SColumnInfoData *pInputData = pInput->columnData;
+  SColumnInfoData *pOutputData = pOutput->columnData;
+  int32_t bufSize = TSDB_MAX_FIELD_LEN + 1;
+  char *buf = taosMemoryMalloc(bufSize);
```
看起来没有必要分配和拷贝到buf中，只需要一个指针就可以了

aestriplex commented:
@dapan1121 I fixed both function.md and sclfunc.c. Also, I rebased my branch on upstream/main (originally I rebased it on upstream 3.3.6, but I noticed that the base branch of this PR has been changed), so there are 6 commits now on top of upstream/head.

![Screenshot 2025-04-10 at 1 36 57 AM](https://github.com/user-attachments/assets/d9854c08-bdcf-4d3d-9a03-49fdeb9c5957)

I am noticing that it's taking a while to updating this PR after the rebase - I'm still seeing the loader with "Processing updates". This is probably due to the rebase. If necessary, I could open another pull request.

aestriplex commented:
This PR still seems to be stuck on “Processing updates.” Probably something went wrong in Github. I'll close it and open another one from the same branch
